### ISSUE TITLE
Identify reasons for memory spikes and crashes

### ISSUE THREAD

**Issue by arpee** [SaitoTech/saito-rust-workspace#712]
![Image](https://github.com/SaitoTech/saito-rust-workspace/assets/2867779/e94a415d-fc76-45ac-b175-76aeb04d8fed)

We continue to have sudden increases in memory use during operations. We need to find a way to identify these so we can log and address them.

These spikes sometimes result in crashes due to out of memory problems.

**Comment by arpee** [comment]
![Image](https://github.com/SaitoTech/saito-rust-workspace/assets/2867779/ef5a6f6d-5da0-4c50-93db-317e48f8b766)

FYI Node Crash.

**Comment by arpee** [comment]
More info from discussion with Sanka in TG.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:04 AM]


Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:05 AM]
So - we had another crash... and I was trying to find the logs... and saw this.

Sanka Darshana, [13/5/2024 10:05 AM]
I'll reduce the logs. I think I added too much logs to find that block fetch issue

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:05 AM]
So the crash happened at 2024-05-12 just before 6am.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:06 AM]
So in a couple of minutes we produced 200MB of logs.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:06 AM]
I actually think that's going to tell us a lot.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:07 AM]
So maybe don't :)

Sanka Darshana, [13/5/2024 10:07 AM]
ok. 😃 I thought you sent the screenshot to say too much logs

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:17 AM]
Not this time.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:17 AM]
     1 2024-05-12T05:51:13
      4 2024-05-12T05:51:16
     37 2024-05-12T05:51:23
    761 2024-05-12T05:51:24
   8345 2024-05-12T05:51:25
      2 2024-05-12T05:51:32
      2 2024-05-12T05:54:41
  47956 2024-05-12T05:54:43
 103251 2024-05-12T05:54:44
 120582 2024-05-12T05:54:45
 122808 2024-05-12T05:54:46
  85694 2024-05-12T05:54:47
  74208 2024-05-12T05:54:48
  74738 2024-05-12T05:54:49
  48935 2024-05-12T05:54:50
  73978 2024-05-12T05:54:51
  64476 2024-05-12T05:54:52
  72551 2024-05-12T05:54:53
  66930 2024-05-12T05:54:54
  75967 2024-05-12T05:54:55
  86627 2024-05-12T05:54:56
  90548 2024-05-12T05:54:57
 111161 2024-05-12T05:54:58
 107058 2024-05-12T05:54:59
  60923 2024-05-12T05:55:00
  63820 2024-05-12T05:55:01
  49558 2024-05-12T05:55:02
  65815 2024-05-12T05:55:03
  77003 2024-05-12T05:55:04
  72822 2024-05-12T05:55:05
  75520 2024-05-12T05:55:06
  88433 2024-05-12T05:55:07
  93899 2024-05-12T05:55:08
  90639 2024-05-12T05:55:09
  53156 2024-05-12T05:55:10
  49152 2024-05-12T05:55:11
  36864 2024-05-12T05:55:12
  40073 2024-05-12T05:55:13
  15223 2024-05-12T05:55:14

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:17 AM]
So this is the number of rows of log output per second... on the node...

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:18 AM]
above that first row it's between 1 and 100 most of the time.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:18 AM]
Looking around that time...

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:18 AM]
I see a reorg start.

Sanka Darshana, [13/5/2024 10:19 AM]
yes. current guess is reorg is causing this

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:20 AM]
      2 2024-05-12T05:42:33
      2 2024-05-12T05:42:46
      2 2024-05-12T05:43:00
      2 2024-05-12T05:43:15
      2 2024-05-12T05:43:59
      2 2024-05-12T05:44:30
      2 2024-05-12T05:44:31
      2 2024-05-12T05:44:44
      2 2024-05-12T05:46:06
    342 2024-05-12T05:51:25
   4492 2024-05-12T05:54:43
   9189 2024-05-12T05:54:44
  10347 2024-05-12T05:54:45
  11300 2024-05-12T05:54:46
   7246 2024-05-12T05:54:47
   6708 2024-05-12T05:54:48
   6725 2024-05-12T05:54:49
   3954 2024-05-12T05:54:50
   6635 2024-05-12T05:54:51
   5640 2024-05-12T05:54:52
   6502 2024-05-12T05:54:53
   5955 2024-05-12T05:54:54
   6845 2024-05-12T05:54:55
   7437 2024-05-12T05:54:56
   8345 2024-05-12T05:54:57
   9616 2024-05-12T05:54:58
   9238 2024-05-12T05:54:59
   5739 2024-05-12T05:55:00
   5739 2024-05-12T05:55:01
   4025 2024-05-12T05:55:02
   5755 2024-05-12T05:55:03
   7010 2024-05-12T05:55:04
   6507 2024-05-12T05:55:05
   6790 2024-05-12T05:55:06
   7538 2024-05-12T05:55:07
   8127 2024-05-12T05:55:08
   8435 2024-05-12T05:55:09
   4400 2024-05-12T05:55:10
   4615 2024-05-12T05:55:11
   3035 2024-05-12T05:55:12
   3277 2024-05-12T05:55:13
   1624 2024-05-12T05:55:14

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 10:20 AM]
So, that's the same thing - but filtering for rows that contain 'reorg'....

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:19 AM]
How are things going BTW.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:20 AM]
I have been leaving you to things as you have been making great progress.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:20 AM]
We should have a 3p meeting with David later this week on next steps - as I get the feeling you'll have some resolution for this fork/reorg stuff soon :)

Sanka Darshana, [13/5/2024 11:23 AM]
work is usual. got a call around 2AM that 2 people from the restaurant got into an accident while drunk driving (motorcycle).

Sanka Darshana, [13/5/2024 11:23 AM]
so life is also... usual

Sanka Darshana, [13/5/2024 11:23 AM]
need to think of a way to easily recreate the issue

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:24 AM]
Oof... everyone ok?

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:24 AM]
Yeah... not sure how we re-create.

Sanka Darshana, [13/5/2024 11:24 AM]
one is in ICU. other's arm is broken.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:25 AM]
Also - should a browser/lite-client be able to trigger a reorg on a full node?

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:25 AM]
Oh, no....

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:25 AM]
I mean, don't drink then get on a motorbike... but, still that's not fun.

Sanka Darshana, [13/5/2024 11:25 AM]
let's see what happens.

Sanka Darshana, [13/5/2024 11:26 AM]
should not. but can't be sure

Sanka Darshana, [13/5/2024 11:26 AM]
there could be some odd bug

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:26 AM]
Right...

Sanka Darshana, [13/5/2024 11:26 AM]
I mean, block fetch issue was not really related to  block fetching but merkle root / lite block generation

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:26 AM]
It's a double problem... lite node should not be able to trigger this... but, also what if a full node has the same kind of chain and connects.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:27 AM]
Is it maybe that the browser has blocks that are just not on the current chain?

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:27 AM]
like we have reset in between and it has not gotten the message?

Sanka Darshana, [13/5/2024 11:27 AM]
could be. but hard to say without recreating it

Sanka Darshana, [13/5/2024 11:28 AM]
even if I have a fix, need to know how to recreate so I know it's fixed after

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:28 AM]
right.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:28 AM]
maybe I can try to bottle it locally...

Sanka Darshana, [13/5/2024 11:29 AM]
I downloaded the prod blocks to my laptop. will test with it too

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:30 AM]
yeah - thinking about how to create a browser that has the wrong chain without it resetting itself...

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:30 AM]
will test that first.

Sanka Darshana, [13/5/2024 11:31 AM]
this should be easy I think. just nuke the server ?

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:32 AM]
yes - plan is to start node locally and connect a browser... get it a few blocks... then restart the node but with the blocks from prod... then reopen the browser...

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 11:45 AM]
well - that didn't trigger anything.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 12:04 PM]
So... i did a few tests without success.

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 12:05 PM]
I tried:
 - copying the blocks from prod
 - starting the node with the first 1000 blocks from prod and letting a clean browser sync them
 - did some stuff creating another 10 blocks or so
 - stopped the node leaving the browser open
 - copied the remaining 60k blocks from prod into the blocks folder and restarted the node
 - watched the browser sync the chain...

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 12:06 PM]
And some variations - no crazy load on my laptop.

Sanka Darshana, [13/5/2024 12:51 PM]
these things take a bit of time. fixing them is quick if I know how to recreate them

Richard Parris 🟥 (Notifications off by default)., [13/5/2024 1:16 PM]
en...

Sanka Darshana, [13/5/2024 1:24 PM]
it probably is a 2 different chains syncing.

Sanka Darshana, [13/5/2024 1:24 PM]
or there are some inefficiencies in re org code which loads a lot of blocks in to memory

Sanka Darshana, [13/5/2024 1:28 PM]
ok. full block count is 9. so not the blocks in memory
